File: app/src/main/java/me/devsaki/hentoid/fragments/ProgressDialogFragment.java
package me.devsaki.hentoid.fragments;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.PluralsRes;
import androidx.fragment.app.DialogFragment;
import androidx.fragment.app.FragmentManager;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

import me.devsaki.hentoid.R;
import me.devsaki.hentoid.databinding.DialogProgressBinding;
import me.devsaki.hentoid.events.ProcessEvent;

/**
 * Generic dialog to report progress
 */
public class ProgressDialogFragment extends DialogFragment {

    private static final String TITLE = "title";
    private static final String PROGRESS_UNIT = "progressUnit";
    private DialogProgressBinding binding = null;

    private String dialogTitle;
    private @PluralsRes
    int progressUnit;

    public static void invoke(
            @NonNull final FragmentManager fragmentManager,
            @NonNull final String title,
            @PluralsRes final int progressUnit) {
        ProgressDialogFragment fragment = new ProgressDialogFragment();

        Bundle args = new Bundle();
        args.putString(TITLE, title);
        args.putInt(PROGRESS_UNIT, progressUnit);
        fragment.setArguments(args);

        fragment.show(fragmentManager, null);
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        if (null == getArguments()) throw new IllegalArgumentException("No arguments found");
        dialogTitle = getArguments().getString(TITLE, "");
        progressUnit = getArguments().getInt(PROGRESS_UNIT, -1);

        EventBus.getDefault().register(this);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedState) {
        binding = DialogProgressBinding.inflate(inflater, container, false);
        setCancelable(false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View rootView, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(rootView, savedInstanceState);
        binding.title.setText(dialogTitle);
        binding.bar.setIndeterminate(true);
    }

    @Override
    public void onDestroyView() {
        EventBus.getDefault().unregister(this);
        super.onDestroyView();
        binding = null;
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onProcessEvent(ProcessEvent event) {
        if (event.processId != R.id.generic_progress) return;

        binding.bar.setMax(event.elementsTotal);
        binding.bar.setIndeterminate(false);
        if (ProcessEvent.EventType.PROGRESS == event.eventType) {
            binding.progress.setText(getString(
                    R.string.generic_progress,
                    event.elementsOK + event.elementsKO, event.elementsTotal,
                    getResources().getQuantityString(progressUnit, event.elementsOK + event.elementsKO)
            ));
            binding.bar.setProgress(event.elementsOK + event.elementsKO);
        } else if (ProcessEvent.EventType.COMPLETE == event.eventType) {
            dismiss();
        }
    }
}

File: app/src/main/java/me/devsaki/hentoid/fragments/ProgressDialogFragment.kt
package me.devsaki.hentoid.fragments

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.PluralsRes
import androidx.fragment.app.DialogFragment
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import me.devsaki.hentoid.R
import me.devsaki.hentoid.databinding.DialogProgressBinding
import me.devsaki.hentoid.events.ProcessEvent
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode

class ProgressDialogFragment : BaseDialogFragment<Nothing>() {

    companion object {
        const val TITLE = "title"
        const val PROGRESS_UNIT = "progressUnit"

        fun invoke(
            activity: FragmentActivity,
            title: String,
            @PluralsRes progressUnit: Int
        ): DialogFragment {
            val args = getArgs(title, progressUnit)
            return invoke(activity, ProgressDialogFragment(), args, isCancelable = false)
        }

        fun invoke(
            fragment: Fragment,
            title: String,
            @PluralsRes progressUnit: Int
        ): DialogFragment {
            val args = getArgs(title, progressUnit)
            return invoke(fragment, ProgressDialogFragment(), args, isCancelable = false)
        }

        private fun getArgs(title: String, @PluralsRes progressUnit: Int): Bundle {
            val args = Bundle()
            args.putString(TITLE, title)
            args.putInt(PROGRESS_UNIT, progressUnit)
            return args
        }
    }

    private var binding: DialogProgressBinding? = null
    private var dialogTitle: String? = null

    @PluralsRes
    private var progressUnit = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val bundle = arguments
        requireNotNull(bundle) { "No arguments found" }
        dialogTitle = bundle.getString(TITLE, "")
        progressUnit = bundle.getInt(PROGRESS_UNIT, -1)
        EventBus.getDefault().register(this)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedState: Bundle?
    ): View? {
        binding = DialogProgressBinding.inflate(inflater, container, false)
        return binding?.root
    }

    override fun onViewCreated(rootView: View, savedInstanceState: Bundle?) {
        super.onViewCreated(rootView, savedInstanceState)
        binding?.apply {
            title.text = dialogTitle
            bar.isIndeterminate = true
        }
    }

    override fun onDestroyView() {
        EventBus.getDefault().unregister(this)
        super.onDestroyView()
        binding = null
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onProcessEvent(event: ProcessEvent) {
        if (event.processId != R.id.generic_progress) return
        binding?.apply {
            bar.max = event.elementsTotal
            bar.isIndeterminate = false
            val nbProcessed = event.elementsOK + event.elementsKO
            if (ProcessEvent.Type.PROGRESS == event.eventType) {
                progress.text = getString(
                    R.string.generic_progress,
                    event.elementsOK + event.elementsKO,
                    event.elementsTotal,
                    resources.getQuantityString(
                        progressUnit,
                        event.elementsOK + event.elementsKO
                    )
                )
                bar.progress = nbProcessed
            }
            if (ProcessEvent.Type.COMPLETE == event.eventType || (nbProcessed > 0 && nbProcessed == event.elementsTotal)) {
                dismissAllowingStateLoss()
            }
        }
    }

    @Subscribe(sticky = true, threadMode = ThreadMode.MAIN)
    fun onProcessStickyEvent(event: ProcessEvent) {
        if (event.processId != R.id.generic_progress) return
        binding?.apply {
            bar.max = event.elementsTotal
            bar.isIndeterminate = false
        }
        EventBus.getDefault().removeStickyEvent(event)
        if (ProcessEvent.Type.COMPLETE == event.eventType) dismissAllowingStateLoss()
    }
}